File: centour-backend/src/llm_interaction/llm_service.py
# llm_service.py
from google import genai
import os
from dotenv import load_dotenv

load_dotenv()

class llm_service:
    def __init__(self):

        load_dotenv()
        
        api_key = os.getenv("GEMINI_API_KEY")
        self.client = genai.Client(api_key=api_key)
        self.model= "gemini-2.5-flash-lite"

    def ask(self, messages):
        """
        messages: [
            {"role": "User", "content": "Hello"},
            {"role": "System", "content": "Hi there!"}
        ]
        """
        if not messages or len(messages) == 0:
            raise ValueError("❌ Empty messages list passed to ask()")

        chat = self.client.chats.create(model=self.model)
        formatted = "\n".join([f"{m['author']}: {m['message']}" for m in messages])
        response = chat.send_message(formatted)

        # Cevabı döndür
        try:
            return response.text
        except:
            return "⚠️ Gemini yanıt oluşturamadı."

    def ask_stream(self, messages):
        """Streaming metodu - Generator döndürür"""
        if not messages or len(messages) == 0:
            raise ValueError("❌ Empty messages list passed to ask_stream()")

        formatted = "\n".join([f"{m['author']}: {m['message']}" for m in messages])
        
        try:
            # Google GenAI streaming API
            for chunk in self.client.models.generate_content_stream(
                model=self.model,
                contents=formatted
            ):
                if chunk.text:
                    yield chunk.text
                    
        except Exception as e:
            print(f"⚠️ Streaming hatası: {str(e)}")
            yield f"⚠️ Streaming hatası: {str(e)}"

llm_service = llm_service()

File: centour-spring-boot/src/main/java/com/centour/llm/LlmService.java
package com.centour.llm;

import com.centour.document.RetrievedChunk;
import com.centour.message.entity.Message;
import com.google.genai.Client;
import com.google.genai.ResponseStream;
import com.google.genai.types.GenerateContentResponse;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.function.Consumer;
import java.util.stream.Collectors;

@Service
public class LlmService {

    private final Client client;
    private static final String MODEL = "gemini-2.5-flash-lite";

    public LlmService(@Value("${GEMINI_API_KEY}") String apiKey) {
        this.client = new Client.Builder().apiKey(apiKey).build();
    }

    public void askStream(List<Message> messages, List<RetrievedChunk> context, Consumer<String> onChunk) {
        if (messages == null || messages.isEmpty()) {
            throw new IllegalArgumentException("Empty messages list");
        }

        StringBuilder prompt = new StringBuilder();
        if (context != null && !context.isEmpty()) {
            prompt.append("Reference material from the user's attached study documents. ")
                    .append("Use it to ground your answer when relevant, and mention the source filename when you do.\n\n");
            for (RetrievedChunk chunk : context) {
                prompt.append("[").append(chunk.filename()).append("]\n")
                        .append(chunk.content()).append("\n\n");
            }
            prompt.append("---\n\n");
        }

        prompt.append(messages.stream()
                .map(m -> m.getAuthor() + ": " + m.getMessage())
                .collect(Collectors.joining("\n")));

        try (ResponseStream<GenerateContentResponse> stream =
                     client.models.generateContentStream(MODEL, prompt.toString(), null)) {
            for (GenerateContentResponse chunk : stream) {
                String text = chunk.text();
                if (text != null && !text.isEmpty()) {
                    onChunk.accept(text);
                }
            }
        } catch (Exception e) {
            throw new RuntimeException("Gemini streaming error: " + e.getMessage(), e);
        }
    }
}
